src/crypto/CHIPCryptoPALPSA.cpp

```diff
@@ -0,0 +1,1747 @@
+{
+    CHIP_ERROR error                      = CHIP_NO_ERROR;
+    psa_status_t status                   = PSA_SUCCESS;
+    const PsaP256KeypairContext & context = to_const_keypair_ctx(mKeypair);
+    const size_t outputSize               = output.Length() == 0 ? output.Capacity() : output.Length();
+    Encoding::BufferWriter bbuf(output, outputSize);
+    uint8_t privateKey[kP256_PrivateKey_Length];
+    size_t privateKeyLength = 0;
+
+    status = psa_export_key(context.key_id, privateKey, sizeof(privateKey), &privateKeyLength);
+    VerifyOrExit(status == PSA_SUCCESS, error = CHIP_ERROR_INTERNAL);
+    VerifyOrExit(privateKeyLength == kP256_PrivateKey_Length, error = CHIP_ERROR_INTERNAL);
+
+    bbuf.Put(mPublicKey, mPublicKey.Length());
+    bbuf.Put(privateKey, privateKeyLength);
+    VerifyOrExit(bbuf.Fit(), error = CHIP_ERROR_BUFFER_TOO_SMALL);
+    error = output.SetLength(bbuf.Needed());
+
+exit:
+    logPsaError(status);
+
+    return error;
+}
+
+CHIP_ERROR P256Keypair::Deserialize(P256SerializedKeypair & input)
+{
+    VerifyOrReturnError(input.Length() == mPublicKey.Length() + kP256_PrivateKey_Length, CHIP_ERROR_INVALID_ARGUMENT);
+
+    CHIP_ERROR error                = CHIP_NO_ERROR;
+    psa_status_t status             = PSA_SUCCESS;
+    psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
+    PsaP256KeypairContext & context = to_keypair_ctx(mKeypair);
+    Encoding::BufferWriter bbuf(mPublicKey, mPublicKey.Length());
+
+    Clear();
+
+    psa_set_key_type(&attributes, PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1));
+    psa_set_key_bits(&attributes, kP256_PrivateKey_Length * 8);
+    psa_set_key_algorithm(&attributes, PSA_ALG_ECDSA(PSA_ALG_SHA_256));
+    psa_set_key_usage_flags(&attributes, PSA_KEY_USAGE_EXPORT | PSA_KEY_USAGE_SIGN_MESSAGE);
+
+    status = psa_import_key(&attributes, input.ConstBytes() + mPublicKey.Length(), kP256_PrivateKey_Length, &context.key_id);
+    VerifyOrExit(status == PSA_SUCCESS, error = CHIP_ERROR_INTERNAL);
+
+    bbuf.Put(input.ConstBytes(), mPublicKey.Length());
+    VerifyOrExit(bbuf.Fit(), error = CHIP_ERROR_NO_MEMORY);
+
+    mInitialized = true;
+
+exit:
+    logPsaError(status);
+
+    return error;
+}
+
+void P256Keypair::Clear()
+{
+    if (mInitialized)
+    {
+        PsaP256KeypairContext & context = to_keypair_ctx(mKeypair);
+        psa_destroy_key(context.key_id);
+        memset(&context, 0, sizeof(context));
+        mInitialized = false;
+    }
+}
+
+P256Keypair::~P256Keypair()
+{
+    Clear();
+}
+
+CHIP_ERROR P256Keypair::NewCertificateSigningRequest(uint8_t * out_csr, size_t & csr_length) const
+{
+    VerifyOrReturnError(out_csr != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(csr_length > 0, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(mInitialized, CHIP_ERROR_WELL_UNINITIALIZED);
+
+    MutableByteSpan csr(out_csr, csr_length);
+    ReturnErrorOnFailure(GenerateCertificateSigningRequest(this, csr));
```

Comment:
Wondering why other versions don't use the same helper function?

Reply:
I guess the helper was added when all other crypto backends were already implemented. Also, if the other APIs provide a native way to produce a CSR, it may be a better idea to use it.